library/src/main/java/com/bumptech/glide/load/resource/bitmap/ImageReader.java

```diff
@@ -68,6 +80,44 @@ public int getImageOrientation() throws IOException {
 
     @Override
     public void stopGrowingBuffers() {}
+
+    /**
+     * Returns a decoded bitmap for the input stream, ensuring that any associated gainmap is
+     * decoded without errors on Android U if it is a valid gainamp.
+     *
+     * <p>Callers should use the method to decode streams with {@link
+     * BitmapFactory#decodeByteArray(byte[], int, int)} on Android U.
+     *
+     * @param bytes for the bitmap to be decoded.
+     * @param options to be applied in the {@link BitmapFactory#decodeStream} call.
+     */
+    @RequiresApi(VERSION_CODES.UPSIDE_DOWN_CAKE)
+    @Nullable
+    private static Bitmap safeDecodeHardwareBitmapWithGainmap(byte[] bytes, Options options) {
+      Bitmap softwareBitmap = null;
+      try {
+        options.inPreferredConfig = Bitmap.Config.ARGB_8888;
+        softwareBitmap =
+            BitmapFactory.decodeByteArray(bytes, /* offset= */ 0, bytes.length, options);
+        if (softwareBitmap == null) {
+          return null;
```

Comment:
This block of code is repeated a bunch, I think you could do something like this instead:

```
    private static Bitmap safeDecodeHardwareBitmapWithGainmap(byte[] bytes, Options options) {
        try {
            options.inPreferredConfig = Bitmap.Config.ARGB_8888;
            Bitmap softwareBitmap =
                    BitmapFactory.decodeByteArray(bytes, /* offset= */ 0, bytes.length, options);
            return safeDecodeBitmapWithGainmapCommon(softwareBitmap);
        } finally {
            options.inPreferredConfig = Config.HARDWARE;
        }
    }

    private static Bitmap safeDecodeBitmapWithGainmapCommon(Bitmap softwareBitmap) {
        if (softwareBitmap == null) {
            return null;
        }
        try {
            Gainmap gainmap = softwareBitmap.getGainmap();
            if (gainmap != null) {
                Bitmap gainmapContents = gainmap.getGainmapContents();
                if (gainmapContents.getConfig() == Bitmap.Config.ALPHA_8) {
                    softwareBitmap.setGainmap(
                            GainmapCopier.convertSingleChannelGainmapToTripleChannelGainmap(gainmap));
                }
            }
            return softwareBitmap.copy(Bitmap.Config.HARDWARE, /* isMutable= */ false);
        } finally {
            softwareBitmap.recycle();
        }
    }
```

Reply:
Thanks for the tip. DOne.